### PR TITLE
Prevent new users being created with an empty password

### LINKED ISSUE #1656: Users can be created with empty password even though `allow_empty_password` is `False`
When a User is created with POST, [documentation](http://developer.couchbase.com/documentation/mobile/1.2/develop/references/sync-gateway/admin-rest-api/user-admin/post-user/index.html) says:

> password | string | Password of the user that will be created. Required, unless the `allow_empty_password` Sync Gateway per-database configuration value is set to true, in which case the password can be omitted.

However, this does not appear to be true and Users can be created with empty passwords _under all conditions_. This also effects the PUT endpoint - a user that's new can be created without a password.

Many ways to reproduce, but here's a simple one:
- SG config is:
  
  ``` json
  {
      "log": ["*"],
      "adminInterface": "0.0.0.0:4985",
      "interface": "0.0.0.0:4984",
      "databases": {}
  }
  ```
  
  Therefore [`allow_empty_password` is `False` as per the docs](http://developer.couchbase.com/documentation/mobile/1.2/develop/guides/sync-gateway/configuring-sync-gateway/config-properties/index.html#database-configuration):
  
  > Whether to allow empty passwords for Couchbase Server authentication. The default is false.
- Create new database.
- Create user without password, just POST `{'name': 'person'}` to `/db/_user/`

There are no errors reported in the log.

Alternatively, if I update the database's config by posting `{'allow_empty_password': False}` to `/db/_config`, this is accepted (although _all_ data, valid or not appears to be accepted), and Users can still be created without password.

Version string is `Couchbase Sync Gateway/1.2.0(79;9df63a5)`

### PR DESCRIPTION
**Pull request by ajres**
Prevent new users being created with an empty password when config 'allow_empty_password' is set to false.

fixes #1656

### CODE REVIEW

File: rest/admin_api_test.go [hunk 1]
```diff
@@ -1060,7 +1060,7 @@ func TestReplicate(t *testing.T) {
 	var rt restTester
 
 	//Initiate synchronous one shot replication
-	assertStatus(t, rt.sendAdminRequest("POST", "/_replicate", `{"source":"http://localhost:4985/db", "target":"http://localhost:4985/db"}`), 500)
+	assertStatus(t, rt.sendAdminRequest("POST", "/_replicate", `{"source":"http://localhost:4985/db", "target":"http://localhost:4985/db"}`), 200)
 
 	//Initiate asyncronous one shot replication
 	assertStatus(t, rt.sendAdminRequest("POST", "/_replicate", `{"source":"http://localhost:4985/db", "target":"http://localhost:4985/db", "async":true}`), 200)
```
Comment (adamcfraser) [inline rest/admin_api_test.go:1167]: I feel like we need a unit test that reproduces the original bug (#1656), to ensure this doesn't get broken again.

File: db/users.go [hunk 2]
```diff
@@ -26,23 +26,25 @@ type PrincipalConfig struct {
...
-		return false, "Anonymous users should not have a password"
+	if p.Name == nil {
+		if p.Password != nil {
+			return false, "Anonymous users should not have a password"
+		} else {
+			return true, ""
+		}
 	}
 
 	if allowEmptyPass {
```
Comment (adamcfraser) [inline db/users.go]: Should this allowEmptyPass only be returning true if password==nil?  Otherwise it seems like we're allowing invalid (one character) passwords on an update in GetPrincipal (when allowReplace is true).

File: db/users.go [hunk 3]
```diff
@@ -77,7 +79,7 @@ func (dbc *DatabaseContext) UpdatePrincipal(newInfo PrincipalConfig, isUser bool
 	var user auth.User
 	authenticator := dbc.Authenticator()
 	if isUser {
-		isValid, reason := newInfo.IsPasswordValid(dbc.AllowEmptyPassword)
+		isValid, reason := newInfo.IsPasswordValid(dbc.AllowEmptyPassword || allowReplace)
```
Comment (adamcfraser) [inline db/users.go]: What is the allowReplace change accomplishing here?  It looks like we are doing the following:
- on an API call, allowReplace="request != POST"
- for config users, allowReplace = isGuest

I'm not clear on how that interacts with 'allowEmptyPass'.

File: rest/admin_api_test.go (hunk 1 above)
Comment (ajres) [inline rest/admin_api_test.go:1167]: Additional test has been added to ensure users can not be created using PUT or POST with an empty password.

File: db/users.go (hunk 2 above)
Comment (ajres) [inline db/users.go]: After running more tests I have modified this to check for 'password==nil', but this has one side effect, I'm not sure if this is a feature or a bug.

I appears that previously if 'allowEmptyPassword == true' and a user was updated with password = "", that this would remove the users password.

With the current update that ability has been removed, if we want to keep that functionality we can test for password length == 0 as well.

File: db/users.go (hunk 3 above)
Comment (ajres) [inline db/users.go]: Yeah, the 'request != POST' in particular was not working as intended, I have reverted this and take a different approach.

File: rest/admin_api_test.go [hunk 4]
```diff
@@ -98,6 +98,46 @@ func TestUserAPI(t *testing.T) {
...
+	response = rt.sendAdminRequest("POST", "/db/_user/", `{"name":"snej", "password":"a"}`)
+	assertStatus(t, response, 400)
+
+	// PUT update a user with a zero character password, should fail
+	response = rt.sendAdminRequest("POST", "/db/_user/", `{"name":"snej", "password":""}`)
+	assertStatus(t, response, 400)
+
+	// PUT update a user with a three character password, should succeed
+	response = rt.sendAdminRequest("POST", "/db/_user/", `{"name":"snej", "password":"abc"}`)
+	assertStatus(t, response, 201)
```
Comment (adamcfraser) [inline rest/admin_api_test.go:163]: Do we have tests for the cases where allowEmptyPassword=true?

File: db/users.go [hunk 5]
```diff
@@ -110,6 +111,12 @@ func (dbc *DatabaseContext) UpdatePrincipal(newInfo PrincipalConfig, isUser bool
 		return
 	}
 
+	isValid, reason := newInfo.IsPasswordValid(true)
```
Comment (adamcfraser) [inline db/users.go]: Are we calling IsPasswordValid for a second time here if it's a new user?

**Comment by adamcfraser** [comment]
I agree that we want to handle the 'update to set password to empty' if allowEmptyPassword=true.

It feels like the overall logic we want is the following - let me know if you disagree:
1. On create, always validate password.
2. On update, validate password if it's non-nil on the incoming request.
